### PR TITLE
=str #19791 subscription timeouts modes respected in prefixAndTail

### PR DESCRIPTION
**Pull request by ktoso**
Somewhat debateable though, please review guys.

Refs #19791

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndTailSpec.scala [hunk 1]
```diff
@@ -124,6 +124,24 @@ class FlowPrefixAndTailSpec extends AkkaSpec {
       tail.to(Sink.fromSubscriber(subscriber)).run()(tightTimeoutMaterializer)
       subscriber.expectSubscriptionAndError().getMessage should ===("Substream Source has not been materialized in 500 milliseconds")
     }
+    "not fail the stream if substream has been not subscribed in time and configured subscription timeout is noop" in assertAllStagesStopped {
```
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndTailSpec.scala]: has not been

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndTailSpec.scala [hunk 2]
```diff
@@ -124,6 +124,24 @@ class FlowPrefixAndTailSpec extends AkkaSpec {
...
+          .withSubscriptionTimeoutSettings(
+            StreamSubscriptionTimeoutSettings(StreamSubscriptionTimeoutTerminationMode.noop, 1.millisecond)))
+
+      val futureSink = newHeadSink
+      val fut = Source(1 to 2).prefixAndTail(1).runWith(futureSink)(tightTimeoutMaterializer)
+      val (takes, tail) = Await.result(fut, 3.seconds)
+      takes should be(Seq(1))
+
+      val subscriber = TestSubscriber.probe[Int]()
+      Thread.sleep(1000)
```
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndTailSpec.scala]: should be enough with 200?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 3]
```diff
@@ -127,9 +127,20 @@ final class PrefixAndTail[T](n: Int) extends GraphStage[FlowShape[T, (immutable.
...
 
     override protected def onTimer(timerKey: Any): Unit = {
-      val timeout = ActorMaterializer.downcast(interpreter.materializer).settings.subscriptionTimeoutSettings.timeout
-      tailSource.timeout(timeout)
-      if (tailSource.isClosed) completeStage()
+      val materializer = ActorMaterializer.downcast(interpreter.materializer)
+      val timeoutSettings = materializer.settings.subscriptionTimeoutSettings
+      val timeout = timeoutSettings.timeout
+
+      println("timerKey = " + timerKey)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: println

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala [hunk 4]
```diff
@@ -127,9 +127,20 @@ final class PrefixAndTail[T](n: Int) extends GraphStage[FlowShape[T, (immutable.
...
+      timeoutSettings.mode match {
+        case StreamSubscriptionTimeoutTerminationMode.CancelTermination ⇒
+          tailSource.timeout(timeout)
+          if (tailSource.isClosed) completeStage()
+        case StreamSubscriptionTimeoutTerminationMode.NoopTermination ⇒
+        // do nothing
+        case StreamSubscriptionTimeoutTerminationMode.WarnTermination ⇒
+          materializer.logger.warning(s"Substream subscription timeout triggered after $timeout in prefixAndTail($n)!")
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: I don't think we need exclamation marks in log messages, it's not so fantastic

**Comment by patriknw** [comment]
LGTM, but I might miss the point of what is debatable

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndTailSpec.scala (hunk 2 above)
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndTailSpec.scala]: :+1:

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala (hunk 3 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: thanks

File: akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala (hunk 4 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala]: right, removed

**Comment by ktoso** [comment]
Now that I look at it again not sure what I meant by debateable... I think that the noop being dangerous.
Should be ready to merge.

**Comment by ktoso** [comment]
Failure was my mistake (not updated assert text), now fixed

**Pull request merged**